homeassistant/components/control4/media_player.py

```diff
@@ -0,0 +1,406 @@
+        self._id_to_parent = id_to_parent
+        self._sources = sources
+        self._is_soft_on = False
+        self._attr_supported_features = (
+            MediaPlayerEntityFeature.PLAY
+            | MediaPlayerEntityFeature.PAUSE
+            | MediaPlayerEntityFeature.STOP
+            | MediaPlayerEntityFeature.VOLUME_MUTE
+            | MediaPlayerEntityFeature.VOLUME_SET
+            | MediaPlayerEntityFeature.VOLUME_STEP
+            | MediaPlayerEntityFeature.TURN_OFF
+            | MediaPlayerEntityFeature.TURN_ON
+            | MediaPlayerEntityFeature.SELECT_SOURCE
+        )
+
+    def _create_api_object(self):
+        """Create a pyControl4 device object.
+
+        This exists so the director token used is always the latest one, without needing to re-init the entire entity.
+        """
+        return C4Room(self.entry_data[CONF_DIRECTOR], self._idx)
+
+    def _get_device_from_variable(self, var: str) -> int | None:
+        current_device = self.coordinator.data[self._idx][var]
+        if current_device == 0:
+            return None
+
+        return current_device
+
+    def _get_current_video_device_id(self) -> int | None:
+        return self._get_device_from_variable(CONTROL4_CURRENT_VIDEO_DEVICE)
+
+    def _get_current_playing_device_id(self) -> int | None:
+        media_info = self._get_media_info()
+        if media_info:
+            if "medSrcDev" in media_info:
+                return media_info["medSrcDev"]
+            if "deviceid" in media_info:
+                return media_info["deviceid"]
+        return 0
+
+    def _get_media_info(self) -> dict | None:
+        """Get the Media Info Dictionary if populated."""
+        media_info = self.coordinator.data[self._idx][CONTROL4_MEDIA_INFO]
+        if "mediainfo" in media_info:
+            return media_info["mediainfo"]
+        return None
+
+    def _get_current_source_state(self) -> str | None:
+        current_source = self._get_current_playing_device_id()
+        while current_source:
+            current_data = self.coordinator.data.get(current_source, None)
+            if current_data:
+                if current_data.get(CONTROL4_PLAYING, None):
+                    return MediaPlayerState.PLAYING
+                if current_data.get(CONTROL4_PAUSED, None):
+                    return MediaPlayerState.PAUSED
+                if current_data.get(CONTROL4_STOPPED, None):
+                    return MediaPlayerState.ON
+            current_source = self._id_to_parent.get(current_source, None)
+        return None
+
+    @property
+    def device_class(self) -> MediaPlayerDeviceClass | None:
+        """Return the class of this entity."""
+        for avail_source in self._sources.values():
+            if _SourceType.VIDEO in avail_source.source_type:
+                return MediaPlayerDeviceClass.TV
+        return MediaPlayerDeviceClass.SPEAKER
+
+    @property
+    def state(self):
+        """Return whether this room is on or off."""
```

Comment:
I guess it's not a room which is on or off? 🤔

Reply:
In control 4 the room is technically marked as on or off of there are no media sources playing